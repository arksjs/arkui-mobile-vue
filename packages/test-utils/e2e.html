<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
