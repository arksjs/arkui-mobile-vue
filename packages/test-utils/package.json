{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
