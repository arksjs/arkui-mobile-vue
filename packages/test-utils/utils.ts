export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
