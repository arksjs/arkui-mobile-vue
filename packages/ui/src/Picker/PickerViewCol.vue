<template>
  <div class="ak-picker-view_col">
    <VirtualList
      class="ak-picker-view_list"
      :data-index="listIndex"
      :ids="list.map(v => v.value)"
      :itemSize="getItemSize"
      @scroll.stop="onListScroll"
    >
      <template #default="{ index }">
        <div
          class="ak-picker-view_item"
          :class="{
            selected: list[index].selected,
            disabled: list[index].disabled
          }"
          :data-index="index"
        >
          {{ list[index].label }}
        </div>
      </template>
    </VirtualList>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { VirtualList } from '../VirtualList'
import type { ColRow } from './types'
import { DEFAULT_ITEM_HEIGHT } from './util'

export default defineComponent({
  name: 'ak-picker-view-col',
  components: { VirtualList },
  props: {
    list: {
      type: Array as PropType<ColRow[]>,
      required: true
    },
    listIndex: {
      type: Number,
      required: true
    },
    onListScroll: {
      type: Function as PropType<(payload: UIEvent) => void>,
      required: true
    }
  },
  setup(props) {
    function getItemSize(index: number) {
      if (props.list.length === 1) {
        return DEFAULT_ITEM_HEIGHT * 5
      }

      if (index === 0 || index >= props.list.length - 1) {
        return DEFAULT_ITEM_HEIGHT * 3
      }

      return DEFAULT_ITEM_HEIGHT
    }

    return {
      getItemSize
    }
  }
})
</script>
