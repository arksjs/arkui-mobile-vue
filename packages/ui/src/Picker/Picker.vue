<template>
  <div class="fx-picker" :class="{ disabled }" ref="root">
    <SelectorField
      :label="fieldLabel"
      :value="fieldValue"
      :disabled="disabled"
      :name="name"
      :placeholder="placeholder"
      @fieldClick="onFieldClick"
    />
    <PickerPopup
      :options="options"
      :fieldNames="fieldNames"
      :modelValue="modelValue"
      :title="placeholder"
      :formatter="formatter"
      :parser="parser"
      v-model:visible="popupVisible"
      v-if="isInitPopup"
      @confirm="onConfirm"
      ref="popup"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { SelectorField } from '../SelectorField'
import PickerPopup from './PickerPopup.vue'
import { pickerEmits, pickerProps, commonProps } from './props'
import { mergeHandlers } from './util'
import { usePicker } from './use-picker'

export default defineComponent({
  name: 'fx-picker',
  components: { SelectorField, PickerPopup },
  props: { ...commonProps, ...pickerProps },
  emits: { ...pickerEmits },
  setup(props, ctx) {
    return {
      ...usePicker(props, ctx, {
        name: 'picker',
        handlers: mergeHandlers({
          formatter: props.formatter,
          parser: props.parser
        })
      })
    }
  }
})
</script>
