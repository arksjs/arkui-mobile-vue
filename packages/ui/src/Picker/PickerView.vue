<template>
  <div class="ta-picker-view" ref="root">
    <div class="ta-picker-view_cols">
      <ViewCol
        v-for="(colItem, listIndex) in cols"
        :key="colItem.key"
        :list="colItem.rows"
        :listIndex="listIndex"
        :onListScroll="onListScroll"
      />
      <Empty :description="locale.pickerEmptyText" v-if="cols.length === 0" />
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, nextTick, inject, shallowRef } from 'vue'
import { Empty } from '../Empty'
import { frameTo, type PropsToEmits } from '../helpers'
import { pickerViewEmits, pickerViewProps } from './props'
import { usePickerView } from '../Picker/use-picker'
import type { PickerHandlers, PickerViewEmits } from './types'
import { useLocale } from '../ConfigProvider/context'
import ViewCol from './PickerViewCol.vue'
import { DEFAULT_ITEM_HEIGHT, mergeHandlers } from './util'

interface ScrollElement extends HTMLElement {
  scrolling?: boolean
  scrollTimer?: number
}

export default defineComponent({
  name: 'ta-picker-view',
  components: { Empty, ViewCol },
  props: {
    ...pickerViewProps
  },
  emits: { ...pickerViewEmits } as PropsToEmits<PickerViewEmits>,
  setup(props, ctx) {
    const { locale } = useLocale()
    const root = shallowRef<HTMLElement | null>(null)
    const handlers = inject<Partial<PickerHandlers>>('taPickerHandlers', {})

    const { getDetail, cols, isCascade, update, updateColSelected, getValuesByRow, onChange } =
      usePickerView(props, ctx, {
        name: 'picker',
        afterUpdate: updatePos,
        handlers: mergeHandlers(
          {
            formatter: props.formatter,
            parser: props.parser
          },
          handlers
        )
      })

    function updatePos() {
      nextTick(() => {
        // 把选择数据展示在选择框内
        const $picker = root.value as HTMLElement

        if ($picker) {
          const $lists = $picker.querySelectorAll(`.ta-picker-view_list`)

          $lists.forEach(($el, index) => {
            const $list = $el as ScrollElement

            if (cols.value[index] && $list.scrolling !== true) {
              const list = cols.value[index]
              let selectIndex = -1

              for (let i = 0, len = list.rows.length; i < len; i++) {
                if (list.rows[i].selected) {
                  selectIndex = i
                  break
                }
              }

              requestAnimationFrame(() => {
                $list.scrollTop = DEFAULT_ITEM_HEIGHT * selectIndex
              })
            }
          })
        }
      })
    }

    function onListScroll(e: Event) {
      const $list = e.currentTarget as ScrollElement

      if ($list.scrolling) {
        return
      }

      clearTimeout($list.scrollTimer)

      const itemHeight = DEFAULT_ITEM_HEIGHT
      const colIndex = parseInt($list.dataset.index as string)
      const list = cols.value[colIndex]
      let current = Math.round($list.scrollTop / itemHeight)
      let oldSelectIndex = 0

      for (let i = 0; i < list.rows.length; i++) {
        if (list.rows[i].selected) {
          oldSelectIndex = i
          break
        }
      }

      let isChange = current !== oldSelectIndex
      if (isChange) {
        while (list.rows[current].disabled && current !== oldSelectIndex) {
          // 处理disabled 不能选的问题
          if (current > oldSelectIndex) {
            current--
          } else {
            current++
          }
        }
      }
      isChange = current !== oldSelectIndex

      const item = list.rows[current]

      if (current * itemHeight === $list.scrollTop && !isChange) {
        // 如果一致 就不需要修正了
      } else {
        $list.scrollTimer = window.setTimeout(() => {
          // $list.scrollTop = current * itemHeight
          $list.scrolling = true

          frameTo({
            from: $list.scrollTop,
            to: current * itemHeight,
            duration: 100,
            progress(res) {
              $list.scrollTop = res.current
            },
            complete() {
              $list.scrolling = false
            }
          })

          if (isChange) {
            if (isCascade.value) {
              update(getValuesByRow(item))
            } else {
              updateColSelected(colIndex, current)
            }

            onChange()
          }
        }, 400)
      }
    }

    return {
      root,
      cols,
      getDetail,
      onListScroll,
      resize: updatePos,
      locale
    }
  }
})
</script>
