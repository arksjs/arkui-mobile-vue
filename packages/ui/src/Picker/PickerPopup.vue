<template>
  <Drawer
    class="fx-picker-popup"
    placement="bottom"
    :visible="visible"
    @visibleStateChange="onVisibleStateChange"
    @cancel="onCancel"
    @confirm="onConfirm"
    @update:visible="onUpdateVisible"
    ref="popup"
  >
    <template #header>
      <NavBar
        class="fx-drawer_header"
        :title="title"
        :leftButtons="[{ text: locale.pickerCancelText, type: 'primary' }]"
        :rightButtons="[{ text: locale.pickerConfirmText, type: 'primary' }]"
        :iconOnly="false"
        @leftButtonClick="onHeaderLeftClick"
        @rightButtonClick="onHeaderRightClick"
      >
      </NavBar>
    </template>
    <PickerView
      ref="viewRef"
      :modelValue="modelValue"
      :options="options"
      :fieldNames="fieldNames"
      :formatter="formatter"
      :parser="parser"
    />
  </Drawer>
</template>

<script lang="ts">
import { defineComponent, inject } from 'vue'
import PickerView from './PickerView.vue'
import { Drawer } from '../Drawer'
import { NavBar } from '../NavBar'
import { usePopupExtend } from '../popup/use-popup'
import { pickerPopupProps, pickerPopupEmits, commonProps } from './props'
import { usePickerPopup } from '../Picker/use-picker'
import { useLocale } from '../ConfigProvider/context'
import type { SelectorDetail } from '../SelectorField/types'
import type { PickerHandlers } from './types'
import { mergeHandlers } from './util'

export default defineComponent({
  name: 'fx-picker-popup',
  components: { PickerView, Drawer, NavBar },
  props: {
    ...pickerPopupProps,
    ...commonProps,
    title: {
      type: String,
      default: ''
    }
  },
  emits: {
    ...pickerPopupEmits
  },
  setup(props, ctx) {
    const { locale } = useLocale()
    const handlers = inject<Partial<PickerHandlers>>('fxPickerHandlers', {})

    const popup = usePopupExtend<SelectorDetail>(ctx)
    const pickerPopup = usePickerPopup(props, ctx, popup, {
      handlers: mergeHandlers(
        {
          formatter: props.formatter,
          parser: props.parser
        },
        handlers
      )
    })

    return {
      ...popup,
      ...pickerPopup,
      locale
    }
  }
})
</script>
