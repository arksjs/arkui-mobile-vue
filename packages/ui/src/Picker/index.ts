import Picker from './Picker.vue'
import PickerPopup from './PickerPopup.vue'
import PickerView from './PickerView.vue'
import { createConfirmHook, createShowPopup } from '../popup/api'
import type { PopupSuccessConfirmArgs } from '../popup/types'
import type { ShowPickerOptions, PickerDetail } from './types'

const showPicker = createShowPopup<ShowPickerOptions, PopupSuccessConfirmArgs<PickerDetail>>({
  apiName: 'showPicker',
  component: PickerPopup,
  createHook: createConfirmHook
})

export { Picker, PickerPopup, PickerView, showPicker }
export default Picker
