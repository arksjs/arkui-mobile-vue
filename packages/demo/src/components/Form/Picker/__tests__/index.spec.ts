import { shallowMount } from '@vue/test-utils'
import { Picker } from '@/Picker'
import { cascadeOptions } from '../data'

describe('Picker', () => {
  test('should render prop set options correctly', () => {
    const wrapper = shallowMount(Picker, {
      props: {
        options: cascadeOptions
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
