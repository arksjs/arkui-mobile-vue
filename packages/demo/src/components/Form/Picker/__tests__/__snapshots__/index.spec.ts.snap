// Vitest Snapshot v1

exports[`Picker > should render prop set options correctly 1`] = `
"<div class=\\"fx-picker\\">
  <selector-field-stub label=\\"\\" name=\\"\\" placeholder=\\"\\" disabled=\\"false\\" value=\\"\\"></selector-field-stub>
  <!--v-if-->
</div>"
`;
